jib-core: incompatible with jackson 2.12.0 (Java 8 date/time type `java.time.Instant` not supported by default)

**Environment**:

- *Jib version:* jib-core 0.16.0
- *Build tool:* Gradle 6.7.1
- *OS:* Mac, Linux

**Description of the issue**:

I'm writing a gradle plugin leveraging jib-core 0.16.0, and while it works fine on most projects, I've found that if gradle's plugin classpath happens to contain a newer version of jackon (specifically 2.12.0), then `Cache#retrieve` throws:

```
org.gradle.internal.UncheckedException: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.Instant` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: java.util.ArrayList[0]->com.google.cloud.tools.jib.cache.LayerEntriesSelector$LayerEntryTemplate["sourceModificationTime"])
	at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:67)
	at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:41)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:107)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.run(ExecuteActionsTaskExecuter.java:555)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	...
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.util.concurrent.ExecutionException: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.Instant` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: java.util.ArrayList[0]->com.google.cloud.tools.jib.cache.LayerEntriesSelector$LayerEntryTemplate["sourceModificationTime"])
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:552)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:513)
	at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:82)
	at com.google.cloud.tools.jib.builder.steps.StepsRunner.realizeFutures(StepsRunner.java:94)
	at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$buildImage$8(StepsRunner.java:346)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	... 1 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.Instant` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: java.util.ArrayList[0]->com.google.cloud.tools.jib.cache.LayerEntriesSelector$LayerEntryTemplate["sourceModificationTime"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1276)
	at com.fasterxml.jackson.databind.ser.impl.UnsupportedTypeSerializer.serialize(UnsupportedTypeSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4485)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3699)
	at com.google.cloud.tools.jib.json.JsonTemplateMapper.writeTo(JsonTemplateMapper.java:187)
	at com.google.cloud.tools.jib.json.JsonTemplateMapper.writeTo(JsonTemplateMapper.java:173)
	at com.google.cloud.tools.jib.hash.Digests.lambda$computeJsonDigest$0(Digests.java:44)
	at com.google.cloud.tools.jib.hash.Digests.computeDigest(Digests.java:102)
	at com.google.cloud.tools.jib.hash.Digests.computeJsonDigest(Digests.java:45)
	at com.google.cloud.tools.jib.cache.LayerEntriesSelector.generateSelector(LayerEntriesSelector.java:169)
	at com.google.cloud.tools.jib.cache.Cache.retrieve(Cache.java:182)
	at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStep.call(BuildAndCacheApplicationLayerStep.java:102)
	at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStep.call(BuildAndCacheApplicationLayerStep.java:38)
	... 6 more
```

It seems like this behaviour change landed in jackson in this commit: https://github.com/FasterXML/jackson-databind/commit/85c9c8544f0c4f01e88241acc1573746df4f755d#diff-4119ee3f674d1d2c64d16e0118a6cad1277bc30849b60970025af744453158e5R312-R320 and it seems the fix is pretty straightforward, `:jib-core` just needs to explicitly register a JavaTime module when setting up its ObjectMapper rather than relying on the old implicit behaviour to serialize & deserialize Instant types. I'm happy to submit a PR for this.

**Expected behavior**:

Ideally jib-core would work with a variety of different jackson versions, but most importantly latest (2.12.0 at time of writing).

**Steps to reproduce**:

  1. Clone GoogleContainerTools/jib
  2. update build.gradle to build against jackson 2.12.0
  3. Run `./gradlew :jib-core:test -i` and see existing tests fail, e.g.

```
com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStepTest > testRun FAILED
    com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.Instant` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: java.util.ArrayList[0]->com.google.cloud.tools.jib.cache.LayerEntriesSelector$LayerEntryTemplate["sourceModificationTime"])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
        at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1276)
        at com.fasterxml.jackson.databind.ser.impl.UnsupportedTypeSerializer.serialize(UnsupportedTypeSerializer.java:35)
        at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
        at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
        at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
        at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
        at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4485)
        at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3699)
        at com.google.cloud.tools.jib.json.JsonTemplateMapper.writeTo(JsonTemplateMapper.java:187)
        at com.google.cloud.tools.jib.json.JsonTemplateMapper.writeTo(JsonTemplateMapper.java:173)
        at com.google.cloud.tools.jib.hash.Digests.lambda$computeJsonDigest$0(Digests.java:44)
        at com.google.cloud.tools.jib.hash.Digests.computeDigest(Digests.java:102)
        at com.google.cloud.tools.jib.hash.Digests.computeJsonDigest(Digests.java:45)
        at com.google.cloud.tools.jib.cache.LayerEntriesSelector.generateSelector(LayerEntriesSelector.java:169)
        at com.google.cloud.tools.jib.cache.Cache.retrieve(Cache.java:206)
        at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStep.call(BuildAndCacheApplicationLayerStep.java:102)
        at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStepTest.buildFakeLayersToCache(BuildAndCacheApplicationLayerStepTest.java:139)
        at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStepTest.testRun(BuildAndCacheApplicationLayerStepTest.java:158)
```

--- a/build.gradle
+++ b/build.gradle
@@ -66,8 +66,9 @@ subprojects {
 
     COMMONS_COMPRESS: 'org.apache.commons:commons-compress:1.20',
     COMMONS_TEXT: 'org.apache.commons:commons-text:1.9',
-    JACKSON_DATABIND: 'com.fasterxml.jackson.core:jackson-databind:2.11.3',
-    JACKSON_DATAFORMAT_YAML: 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3',
+    JACKSON_DATABIND: 'com.fasterxml.jackson.core:jackson-databind:2.12.0',
+    JACKSON_DATAFORMAT_YAML: 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.12.0',
+    JACKSON_DATATYPE_JSR310: 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.12.0',
     ASM: 'org.ow2.asm:asm:9.0',
     PICOCLI: 'info.picocli:picocli:4.5.2',
 

--- a/jib-core/build.gradle
+++ b/jib-core/build.gradle
@@ -13,6 +13,7 @@ dependencies {
   implementation dependencyStrings.COMMONS_COMPRESS
   implementation dependencyStrings.GUAVA
   implementation dependencyStrings.JACKSON_DATABIND
+  implementation dependencyStrings.JACKSON_DATATYPE_JSR310
   implementation dependencyStrings.ASM
 
   testImplementation dependencyStrings.JUNIT

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/json/JsonTemplateMapper.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/json/JsonTemplateMapper.java
@@ -18,6 +18,7 @@ package com.google.cloud.tools.jib.json;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.type.CollectionType;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
@@ -53,7 +54,8 @@ import java.util.List;
  */
 public class JsonTemplateMapper {
 
-  private static final ObjectMapper objectMapper = new ObjectMapper();
+  private static final ObjectMapper objectMapper =
+      new ObjectMapper().registerModule(new JavaTimeModule());
 
   /**
    * Deserializes a JSON file via a JSON object template.

